site/content/entry/improvements-to-hardware-renderer.md

```diff
@@ -85,11 +87,47 @@ graph: performance of low end gpus
+
+#### Multiplication shouldn't be this slow
+
+When converting from PICA200 shaders into GLSL, there's a few PICA200 opcodes that should just match up without any issues. Addition, subtraction, and multiplication should ... wait. Where did this issue come from?
+
+picture: accurate aot
+subtext: On: I don't see any issue...
+picture: inaccurate aot
+subtext: Off: Oh. Well maybe Eren looks better this way
+
+It turns out that the PICA200 multiplication opcode has a few edge cases that doesn't impact a large majority of games, but leads to some hilarious results in others. On the PICA200, `infinity * 0 = 0` and OpenGL doesn't have a simple way to emulate this behavior. In the generated GLSL shaders, we emulate this behavior by making a function call instead of a simple multiplication.
+
+```glsl
+vec4 sanitize_mul(vec4 lhs, vec4 rhs) {
+    vec4 product = lhs * rhs;
+    return mix(product, mix(mix(vec4(0.0), product, isnan(rhs)), product, isnan(lhs)), isnan(product));
+}
+```
+
+Alas, its a performance penalty to use a function everywhere instead of regular multiplication. On weaker GPUs, we noticed the penalty is so severe, we actually made this configurable. The whole point of a hardware renderer is to be fast, so eating a penalty when only a small handful of games need this level of accuracy would be regrettable. You can turn off this feature in the settings by deselecting "Accurate Shader" and get a noticeable performance boost, but be aware that a few games will break in strange ways!
+
+#### It's a bug, but not where we thought
+
+We were very excited to launch this feature when phantom declared that it was ready; results from user testing was entirely positive, and the performance improvements were unbelievable, but one thing stood in the way. No one had yet tested to see if it worked on AMD GPUs. We called for our good friend [JMC4789]() to break out the AMD card he uses for testing Dolphin, and Citra crashed the driver! Oh no!
+
+From JMC4789's time in Dolphin, he's made a few friends here and there, and so he reached out to a contact that works on AMD drivers. After a few grueling weeks, the driver developer was able to narrow down what the problem is, and surprise, its not a bug in citra, nor it is a bug in the AMD drivers. Its a bug in the GL specification, or more precisely, the issue is ambiguous wording. [glDrawRangeElementsBaseVertex](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glDrawRangeElementsBaseVertex.xhtml) states that the indices should be a pointer, but doesn't say whether the pointer should be to CPU memory or GPU memory. Citra passed a pointer to CPU memory which both Nvidia and Intel drivers were fine with, but AMD drivers refused and crashed instead. As a workaround, phantom added support for streaming storage buffers, which lets Citra work with the data on the CPU and flush it to the GPU when its time to draw.
```

Comment:
I'd rather not say I directly contacted the team.  I think a better way to say it would be that I was able to use connections in Dolphin.  I could ask the person I was contacting if they'd mind being mentioned directly as well.

Also, from what I've learned, what Citra was doing was iffy at best, so, calling out AMD on this is not really the right way to go.  I think this is a case where NVIDIA errs on the side of permissive, and AMD errs on teh side of restrictive.